common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala

```diff
@@ -38,15 +47,157 @@ trait MessageConsumer {
+ * to this actor which will then attempt to fill the pipeline with new messages.
+ *
+ * The actor tries to fill the pipeline with additional messages while the number
+ * of outstanding requests is below the pipeline fill threshold.
+ */
+@throws[IllegalArgumentException]
+class MessageFeed(
+    description: String,
+    logging: Logging,
+    consumer: MessageConsumer,
+    maxPipelineDepth: Int,
+    longPollDuration: FiniteDuration,
+    handler: Array[Byte] => Future[Unit],
+    autoStart: Boolean = true)
+    extends FSM[MessageFeed.FeedState, MessageFeed.FeedData] {
+    import MessageFeed._
+
+    require(consumer.maxPeek <= maxPipelineDepth, "consumer may not yield more messages per peek than permitted by max depth")
+
+    private val pipelineFillThreshold = maxPipelineDepth - consumer.maxPeek
+    private var pipelineOccupancy = 0
+    private implicit val tid = TransactionId.dispatcher
+
+    startWith(Uninitialized, MessageFeed.NoData)
+
+    onTransition {
+        case _ -> FillingPipeline => fillPipeline()
+    }
+
+    when(Uninitialized) {
+        case Event(Ready, _) =>
+            goto(FillingPipeline)
+
+        case _ => stay
+    }
+
+    // fill pipeline either periodically due to a scheduled alarm or when
+    // previously queued messages are released
+    when(FillingPipeline, longPollDuration) {
+        case Event(Processed, _) =>
+            updateOccupancy()
+            if (haveCapacity) {
+                fillPipeline()
+                stay
+            } else {
+                goto(DrainingPipeline)
+            }
+
+        case Event(StateTimeout, _) =>
+            if (haveCapacity) {
+                fillPipeline()
+                stay
+            } else {
+                goto(DrainingPipeline)
+            }
+    }
+
+    when(DrainingPipeline) {
+        case Event(Processed, _) =>
+            // fill if there is room otherwise stay
+            updateOccupancy()
+            if (haveCapacity) {
+                goto(FillingPipeline)
+            } else stay
+
+        case _ => stay
+    }
+
+    if (autoStart) self ! Ready
+
+    private def fillPipeline(): Unit = {
+        if (pipelineOccupancy <= pipelineFillThreshold) {
+            Try {
+                // Grab next batch of messages and commit offsets immediately
+                // essentially marking the activation as having satisfied "at most once"
+                // semantics (this is the point at which the activation is considered started).
+                // If the commit fails, then messages peeked are peeked again on the next poll.
+                // While the commit is synchronous and will block until it completes, at steady
+                // state with enough buffering (i.e., maxPipelineDepth > maxPeek), the latency
+                // of the commit should be masked.
+                val records = consumer.peek(longPollDuration)
+                consumer.commit()
```

Comment:
Are these blocking operations? They'll block the actor and prevent it from receiving potentially state-changing messages. Is that expected?